org.osgi.test.junit5/src/main/java/org/osgi/test/junit5/context/BundleContextExtension.java

```diff
@@ -58,6 +61,12 @@ public void afterEach(ExtensionContext extensionContext) throws Exception {
 		if (closeableResourceBundleContext != null) {
 			closeableResourceBundleContext.close();
 		}
+		bundleContextThreadLocal.remove();
+	}
+
+	@Override
+	public BundleContext getBundleContext() {
+		return bundleContextThreadLocal.get();
```

Comment:
This seems sketchy. How can you ensure this is called in the same thread as the setter? Or a thread that never set.

Reply:
if you have a suggestion let's hear it!

Reply:
to be clear the `beforeEach` method is called in the same thread. that's defined in the contract of the Extension.

Reply:
https://junit.org/junit5/docs/current/api/org/junit/jupiter/api/extension/BeforeEachCallback.html

Reply:
I cannot see how to interpret "before each test is invoked" other than to mean "before, in the same thread, each test is invoked"

Reply:
Why does it have to be thread local? I thought we should not directly hold state? We put it in the extension context. A threadlocal is state. It is just state distributed across threads. And since getBundleContext() here is a public method, anyone can call it from any thread.

Reply:
explain to me how I'm supposed to implement this method otherwise?

Reply:
This stateless business doesn't make any real sense to be honest. I can see it being relevant only when using
```
@ExtendWith(FooExtension.class)
public class Tests {...}
```
Because there's certainly no other way to handle state.

However, the Junit5 documentation shows instance Extensions being used exactly like above:
https://junit.org/junit5/docs/current/user-guide/#extensions-registration-programmatic-instance-fields

Reply:
I don't know why you add the method and who is supposed to use it. But if you are going to store state, a threadlocal is not the place to do it since only one thread can see the state.

Reply:
I changed it to not use a TL.

Reply:
it's used to install bundles from the instance of the extension during a setup method or during the test method itself!

Reply:
more specifically, it's used to implement the `InstallBundle` interface.